Project.ipynb

```diff
@@ -149,10 +153,64 @@
-    "def add_temporal_dim(A,n_updates):\n",
+    "## disable some links for a certain amount of time\n",
+    "##\n",
+    "## modes:\n",
+    "## light mode: few links, short time\n",
+    "## heavy mode: few links, long time\n",
+    "## unstable mode: many links, short time\n",
+    "## extreme mode: many links, long time\n",
+    "##\n",
+    "## priority: random, near, far\n",
+    "## prioritize disabling of links involving nodes near or far from Earth\n",
+    "## \n",
+    "## custom_n_targets = number of targets\n",
+    "## custom_offtimes: array containing possible values for off times expressed in number of updates of A\n",
+    "\n",
+    "def disable_links(A,n_updates, mode=\"light\",priority=\"random\", custom_n_targets = None, custom_offtimes = None):\n",
+    "\n",
+    "    arrays = [A for i in range(n_updates)]\n",
+    "    At = np.stack(arrays,axis=2)      #adds temporal dimension\n",
+    "    \n",
+    "    t_nodes = np.arange(n)\n",
+    "    \n",
+    "    if priority == \"random\": np.random.shuffle(t_nodes)   #nodes targeted for link removal\n",
+    "    elif priority == \"near\": t_nodes = sorted_nodes[:int(np.ceil(n/3))]\n",
+    "    elif priority == \"far\": t_nodes = np.flip(sorted_nodes)[:int(np.ceil(n/3))]\n",
     "    \n",
-    "    At = np.broadcast_to(A.reshape((n,n,1)),(n,n,n_updates))\n",
-    "    ### At = ???(At) ##rimuovere link disattivati\n",
+    "    possible_links = []\n",
+    "    possible_links = np.array([[x,y] for x in t_nodes for y in range(n) if x!=y if A[x,y]<tau_max if [y,x] not in possible_links])\n",
+    "\n",
+    "    if mode == \"light\":  \n",
+    "        n_targets = possible_links.shape[0]//3\n",
+    "        offtimes = [n_updates//10]   # selected links stay off for around 1/10 of the total time\n",
+    "            \n",
+    "    if mode == \"heavy\":\n",
+    "        n_targets = possible_links.shape[0]//3\n",
+    "        offtimes = [n_updates//3]\n",
+    "        \n",
+    "    if mode == \"unstable\":\n",
+    "        n_targets = possible_links.shape[0]*3//4\n",
+    "        offtimes = [n_updates//10]\n",
+    "        \n",
+    "    if mode == \"extreme\":\n",
+    "        n_targets = possible_links.shape[0]*3//4\n",
+    "        offtimes = [n_updates//3]\n",
+    "        \n",
+    "    if custom_n_targets != None: n_targets = custom_n\n",
+    "    if custom_offtimes != None: offtimes = custom_offtimes\n",
+    " \n",
+    "    disabled_links = np.random.permutation(possible_links)[:n_targets,:]\n",
+    "    \n",
+    "    for link in disabled_links:\n",
+    "        print(link)\n",
+    "        start = np.random.randint(0,n_updates)\n",
+    "        print(\"start {}\".format(start))\n",
+    "        end = start + np.random.choice(offtimes)\n",
+    "        print(\"end {}\".format(end))\n",
+    "        for t in range(start,min(n_updates,end)):\n",
+    "            At[link[0],link[1],t] = At[link[1],link[0],t] = np.inf\n",
+    "            \n",
```

Comment:
qui si dovrebbe riuscire a togliere il ciclo for facendo una cosa del tipo 
At[link[0], link[1], start, min(n_updates, end)] = np.inf